azurerm/internal/services/timeseriesinsights/tests/resource_time_series_insights_environment_test.go

```diff
@@ -0,0 +1,167 @@
+					testCheckAzureRMTimeSeriesInsightsEnvironmentExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+		},
+	})
+}
+
+func TestAccAzureRMTimeSeriesInsightsEnvironment_complete(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_time_series_insights_environment", "test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: testCheckAzureRMTimeSeriesInsightsEnvironmentDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAzureRMTimeSeriesInsightsEnvironment_basic(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMTimeSeriesInsightsEnvironmentExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+			{
+				Config: testAccAzureRMTimeSeriesInsightsEnvironment_complete(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMTimeSeriesInsightsEnvironmentExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+			{
+				Config: testAccAzureRMTimeSeriesInsightsEnvironment_basic(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMTimeSeriesInsightsEnvironmentExists(data.ResourceName),
+				),
+			},
+			data.ImportStep(),
+		},
+	})
+}
+
+func testCheckAzureRMTimeSeriesInsightsEnvironmentExists(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		client := acceptance.AzureProvider.Meta().(*clients.Client).TimeSeriesInsights.EnvironmentsClient
+		ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+		// Ensure we have enough information in state to look up in API
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		id, err := parse.TimeSeriesInsightsEnvironmentID(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		resp, err := client.Get(ctx, id.ResourceGroup, id.Name, "")
+		if err != nil {
+			return fmt.Errorf("Bad: Get on TimeSeriesInsightsEnvironmentClient: %+v", err)
+		}
+
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Bad: Time Series Insights Environment %q (resource group: %q) does not exist", id.Name, id.ResourceGroup)
+		}
+
+		return nil
+	}
+}
+
+func testCheckAzureRMTimeSeriesInsightsEnvironmentDestroy(s *terraform.State) error {
+	client := acceptance.AzureProvider.Meta().(*clients.Client).TimeSeriesInsights.EnvironmentsClient
+	ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "azurerm_time_series_insights_environment" {
+			continue
+		}
+
+		id, err := parse.TimeSeriesInsightsEnvironmentID(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+		resp, err := client.Get(ctx, id.ResourceGroup, id.Name, "")
+
+		if err != nil {
+			return nil
+		}
+
+		if resp.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("Time Series Insights Environment still exists: %q", id.Name)
+		}
```

Comment:
This reads really weird. If the resource does not exist, the error would be non-nil, and this check would return `nil` immediately. But if something is wrong (for instance 429 returned), the error would be also non-nil. In this case, this case would regard this as "the resource does not exist" and return nil.

On the other hand, the `if resp.StatusCode != http.StatusNotFound` would never succeed. `Get` function would only return nil error when it gets the resource successfully (in other word `200`), in this case the status code would never be 404.
This behaviour is also defined in the swagger [here](https://github.com/Azure/azure-rest-api-specs/blob/b069b0fa4817788d6363007645cae04317864f09/specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/preview/2018-08-15-preview/timeseriesinsights.json#L162)

Reply:
And to be honest, a lot of the existing tests have this kind of check destroy functions, which I suppose it is not so correct. What do you think?